Data_Manipulation/voter_participation_crosswalk.R

```diff
@@ -0,0 +1,229 @@
+  pop_of_intersection <- aw_interpolate(zctas_overlapping_polys,
+                                        tid = precinct_zctas,
+                                        source =  blocks_sf,
+                                        sid = block_id,
+                                        weight = "sum",
+                                        output = "sf",
+                                        extensive = "pop10") %>%
+    mutate(pop10= round(pop10,0))  %>%
+    sf::st_make_valid() %>%
+    na.omit()
+  if(sum(pop_of_intersection$pop10) != sum(blocks_sf$pop10)) warning(paste0("Not all the population has been attributed to the state. ",sum(pop_of_intersection$pop10), " is the sum of the population that has been attributed and ", sum(blocks_sf$pop10)," is the total population from the census file"))
+  
+  
+  #### Step 4: Create a dataframe showing the percent population  #### 
+  crosswalk <- pop_of_intersection %>%
+    separate(precinct_zctas, c("precinct", "zcta"), sep = "_") %>%
+    group_by(precinct, zcta) %>%
+    summarise(pop10 = sum(pop10)) %>%
+    group_by(precinct) %>%
+    mutate(frac_of_precinct = {if(sum(pop10) > 0) pop10/sum(pop10) else 1/n()}) %>%
+    group_by(zcta) %>%
+    mutate(frac_of_zctas = {if(sum(pop10) > 0) pop10/sum(pop10) else 1/n()}) %>%
+    group_by()
+  
+  #### Step 5: Apportion the votes by population weighting for each ZCTA  #### 
+  general_zctas_crosswalked <- as.data.frame(vote_sf) %>%
+    left_join(crosswalk, by = c("geo_id"= "precinct")) %>%
+    group_by(zcta) %>%
+    summarise(allocated_presidential_votes = round(sum(presidential_votes * frac_of_precinct)),0) %>%
+    as.data.frame() %>%
+    mutate(zcta = str_sub(zcta, start = -5))
+  
+  
+  #### Step 6: Add the denominator - 18+ population to the final file  #### 
+  final_df <- left_join(general_zctas_crosswalked,over_18_pop, by = "zcta") %>%
+    as.data.frame() %>%
+    mutate(voter_percent_of_population = ifelse((allocated_presidential_votes/pop19)>1,1,allocated_presidential_votes/pop19)) %>%
+    select(zcta,pop19,allocated_presidential_votes,voter_percent_of_population) %>%
+    rename(adult_population = pop19,
+           presidential_votes = allocated_presidential_votes)
+  
+  if(length(unique(general_zctas_crosswalked$zctas)) != length(unique(general_zctas_crosswalked$zctas_id))) warning(paste0("Some ZCTAs have been dropped from the final data frame. ",length(unique(zctas_sf$zctas)), " is the number of unique zctas in the original shpae files and ", length(unique(final_df$zctas))," is the number of unique zctas in the final dataframe."))
+  
+  # write each state file out to the prepossessed folder to save a copy 
+  write.csv(final_df, 
+            file = file.path(
+              paste0(
+              data_folder,
+              "/",state_abv,"_voter_participation.csv"
+            ), 
+            row.names = F, 
+            na = ""))
```

Comment:
Please create a subdirectory in the data_folder using `dir.create()`. Then verify that all 50 states were successful below, rbind them, and delete the directory (rather than doing it manually after the fact).